proto/pocket/service/relay.proto

```diff
@@ -24,82 +24,23 @@ message RelayRequestMetadata {
 // RelayRequest holds the request details for a relay.
 message RelayRequest {
     RelayRequestMetadata meta = 1;
-    oneof payload {
-        JSONRPCRequestPayload json_rpc_payload = 2;
-        RESTRequestPayload rest_payload = 3;
-        // Future possible relay types:
-        // WebSocketsRequestPayload websockets_payload = 4;
-        // GRPCRequestPayload grpc_payload = 5;
-        // GraphQLRequestPayload graphql_payload = 6;
-    }
-}
-
-// JSONRPCRequestPayload contains the payload for a JSON-RPC request.
-// See https://www.jsonrpc.org/specification#request_object for more details.
-message JSONRPCRequestPayload {
-    uint32 id = 1; // Identifier established by the Client to create context for the request.
-    string jsonrpc = 2; // Version of JSON-RPC. Must be exactly "2.0".
-    string method = 3; // Method being invoked on the server.
-    // TODO_TECHDEBT(#126): Make params a `oneof` of a list or map per the JSON-RPC specifications
-    // should they be a list of maps?
-    //map<string, string> params = 4; // Parameters for the method. https://www.jsonrpc.org/specification#parameter_structures
-    repeated string params = 4; // Parameters for the method. https://www.jsonrpc.org/specification#parameter_structures
-}
-
-// RESTRequestType represents the type of REST request.
-enum RESTRequestType {
-    REST_REQUEST_TYPE_UNKNOWN = 0; // Default uninitialized value.
-    REST_REQUEST_TYPE_GET = 1;
-    REST_REQUEST_TYPE_PUT = 2;
-    REST_REQUEST_TYPE_POST = 3;
-    REST_REQUEST_TYPE_DELETE = 4;
-}
-
-// RESTRequestPayload contains the payload and metadata for a REST request.
-message RESTRequestPayload {
-    RESTRequestType request_type = 1;
-    string http_path = 2; // Path for the REST endpoint.
-    string contents = 3; // Request contents.
-    map<string, string> headers = 4; // Request headers.
+    // payload is the serialised payload for the request.
+    // The payload is passed directly to the service and as such can be any
+    // format that the service supports: JSON-RPC, REST, gRPC, etc.
+    bytes payload = 2;
 }
 
 // RelayResponse contains the response details for a RelayRequest.
 message RelayResponse {
     RelayResponseMetadata meta = 1;
-    oneof payload {
-        JSONRPCResponsePayload json_rpc_payload = 2;
-        RESTResponsePayload rest_payload = 3;
-        // Future possible relay types:
-        // WebSocketsResponsePayload websockets_payload = 4;
-        // GRPCResponsePayload grpc_payload = 5;
-        // GraphQLResponsePayload graphql_payload = 6;
-    }
+    // payload is the serialised payload for the response.
+    // The payload is passed directly from the service and as such can be any
+    // format the the service responds with: JSON-RPC, REST, gRPC, etc.
```

Comment:
@h5law , @Olshansk , @bryanchriswhite
I think this raises very pertinent questions regarding how we categorize our services.

Except the need to know the RPC type for:
- Accounting: Know the methods of the JSON-RPC calls and HTTP headers and paths for REST...
- Internal error replies: e.g. We reply from our internal errors with a 500 status code in the header for HTTP, stick an Id to a JSON-RPC reply for this one...

But workflow-wise we could categorize them into:
- Symmetric protocols: To each request corresponds a reply (1-1). These almost exclusively rely on TCP/HTTP transport (? At least we'll be sticking to it for a some time). This could include JSON-RPC, REST and gRPC
- Asymmetric protocols: Where we could have many replies for a single request (e.g. subscriptions) or many requests for a single reply (e.g. chunked file uploads) these are n-m req/res protocols. They may transport over TCP (messaging web apps...) or UDP/WebRTC (video streaming, gaming, etc...). This could include Websocket, gRPC and JSON-RPC

If we adhere to just passing the bytes of the original request, it doesn't matter which RPC type it is but knowing if we're gonna handle stateful or stateless communication matters a lot and will dictate which part of the code is going to handle that request.

Our current jsonrpc.go servers could support all the symmetric protocols with very low effort and both the appgateserver and the relayerproxy could be agnostic of the RPC type as long as its symmetric.

We could then implement the asymmetric ones (starting with Websockets then gRPC) but it would be the same logic/flow handling them IMO.

We could couple that to partial decoders given the RPC type is known (regardless of its flow) to extract the relevant info. for accounting and error handling.

WDYT?

Reply:
Will read this through more later and come back with a full response @red-0ne however, we could look into adding some specifics to the config file:

As an application I want to send json-rpc requests to anvil over a websocket connection (this would be in the app's config.yaml file on staking)

The protocol can then read this and treat any incoming requests according to this on chain service config on behalf of the application - avoiding reading interim unmarshal steps. This would enforce correct config setups as we would start erroring on invalid configs but thats a user error.

Will come back with an actual response sooon ™️

Reply:
@red-0ne Thanks for taking the initiative for asking the right questions. There's A LOT to what you asked and I think it will expand much further than just a github thread. 

Wdyt of starting a discussion [1] with what you shared above? We can use it as a starting point to discuss some of our goals, user requirements, features, etc, and then turn that into a proper ADR or design doc before moving forward with more concrete features?

I also believe the spike with the backend team will help shed some light on it.

[1] https://github.com/pokt-network/poktroll/discussions

Reply:
@red-0ne Bump

Reply:
Created a discussion for it https://github.com/pokt-network/poktroll/discussions/207

Reply:
@red-0ne As discussed offline, few requests for the discussion:

1. Please update formatting
2. Please reference the PRs harry has put together (as a starting point)

We want to make it easy for others to understand and join in on the discussion.